### ISSUE TITLE
`__tostring` produces "random" pointers

### ISSUE THREAD

**Issue by psychon** [way-cooler/way-cooler#507]
```lua
first, second = button{}, button{}
print(first)
print(second)
```
results for me in
```
button: 0x7f495bb449e0
button: 0x7f495bb449e0
```
As you see, this is two times the same pointer for different buttons.

If you want me to guess, the problem is in `awesome::object::default_tostring`. This uses `&foo as *const _` on the passed-in `AnyUserData` instance. However, that one is stack-allocated (right?) and so we just get a random stack pointer here.

**Comment by psychon** [comment]
Hm, I guess this is (finally) a real use case for https://github.com/chucklefish/rlua/issues/76.
This needs to be fixed in `default_tostring` in `src/awesome/object.rs`. This code does not know the concrete type of object, so cannot use `borrow` to get a pointer. No idea how to get a sane pointer here.

**Comment by Timidger** [comment]
Yeah I don't think it's possible to get a pointer if you don't know the value. Though maybe try using `Any` for `borrow` first?

**Comment by psychon** [comment]
I can try that (just not right now), but I doubt that it works. rlua uses a [`HashMap<TypeId, c_int>`](https://github.com/chucklefish/rlua/blob/8b9ab3d0311dfe2cc52453f429843d2cabe71282/src/lua.rs#L1291) to [save](https://github.com/chucklefish/rlua/blob/8b9ab3d0311dfe2cc52453f429843d2cabe71282/src/lua.rs#L1001) and [retrieve](https://github.com/chucklefish/rlua/blob/8b9ab3d0311dfe2cc52453f429843d2cabe71282/src/lua.rs#L897) the (unique) metatable for a type. Since this is a map, `TypeId::of::<Any>` will not produce the same entry as `TypeId::of::<ButtonState>`. :-/

However, perhaps it is a good idea to add `Any`-support to `rlua::AnyUserData`. That would require adding something like `borrow_any` and `borrow_mut_any` which return a reference to `Any`. However, that would mean that the `UserData` actually has to implement `Any`, which is a requirement that so far does not exist (right?).

**Comment by psychon** [comment]
Oh, does not even compile:
```
error: the `borrow` method cannot be invoked on a trait object
   --> src/awesome/object.rs:291:21
    |
291 |     let ptr = &*obj.borrow::<Any>()?;
    |                     ^^^^^^
```
